### ISSUE TITLE
str() call won't accept char * arguments

### ISSUE THREAD

**Issue by brad-lewis** [iovisor/bpftrace#1010]
I get the following error after inserting a call to str() in an attempt to reduce the size used  a char[256] from a structure.

```
: ERROR: str() only supports integer arguments (string provided)

        $spa_name = str((((struct zfsvfs *) ((struct inode *)arg0)->i_sb->s_fs_info)->z_os->os_spa->spa_name));
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

From the description in the reference guide this seems like it should be a valid call:
```
str(): Strings
Syntax: str(char *s [, int length])

Returns the string pointed to by s. length can be used to limit the size of the read, and/or introduce a null-terminator. By default, the string will have size 64 bytes (tuneable using env var BPFTRACE_STRLEN).
```

Again, the reason is for inserting the str() call is to try and avoid the stack overflow caused by this probe function. 

```
kprobe:zfs_read,kprobe:zfs_write
{
	$spa_name = ((struct zfsvfs *) ((struct inode *)arg0)->i_sb->s_fs_info)->z_os->os_spa->spa_name;
	@zfs_count[$spa_name] = count();
}
```

**Comment by mmisono** [comment]
I don't try it, but I guess the following workaround will work.

```
struct my_spa_t {
  char spa_name[64];
  // define other fields if needed
}

kprobe:zfs_read,kprobe:zfs_write
{
  $spa = (struct my_spa_t*)(((struct zfsvfs *)((struct inode *)arg0)->i_sb->s_fs_info)->z_os->os_spa);
  zfs_count[$spa->$spa_name] = count();
}
```

**Comment by brad-lewis** [comment]
Thanks @mmisono.   Thanks for the workaround.  Since spa_name is the first field in spa_t and is likely to remain the first field this may work pretty well.  I have seen other cases similar cases where this becomes and issue.  Would it be possible to provide a solution through str() or someway to truncate strings like this?   Ideally truncate as it is copied to a key so it only needs to be stored on the stack once.

**Comment by brendangregg** [comment]
What's the reason for the stack overflow? Is it the string itself? (Check with -v if it's not already printing out the bytecode.) We are planing to move strings from stack storage to map storage, and there was a draft PR for doing this; it will provide unlimited size strings.

As for spa_name -- it's a char array, not a char *. https://github.com/zfsonlinux/zfs/blob/e64e84eca57966cff1314d61f5b9557876e856e3/include/sys/spa_impl.h#L204

```C
	char		spa_name[ZFS_MAX_DATASET_NAME_LEN];	/* pool name */
```

It sounds like you're doing this:

```
# bpftrace --include linux/sched.h -e 'k:do_nanosleep { printf("%s\n", str(((struct task_struct *)curtask)->comm)); }'
stdin:1:33-76: ERROR: str() only supports integer arguments (string provided)
```

When you could be doing this:

```
# bpftrace --include linux/sched.h -e 'k:do_nanosleep { printf("%s\n", ((struct task_struct *)curtask)->comm); }'
Attaching 1 probe...
ModuleProcessTh
ModuleProcessTh
SendControllerT
[...]
```

This also works:

```
# bpftrace --include linux/sched.h -e 'k:do_nanosleep { $name = ((struct task_struct *)curtask)->comm; printf("%s\n", $name); }'
Attaching 1 probe...
SendControllerT
SendControllerT
[...]
```

If you need truncation functionality, then we can add it: there's already some string.h functions in the works.

**Comment by mmisono** [comment]
bpftrace treats `char *` and `char []` differently, and the type of former is `Type::integer` and the latter is `Type::string`. When accessing a struct field whose type is `Type::string` (e.g., `$spa->spa_name`), bpftrace first copies the whole buffer to the stack.

https://github.com/iovisor/bpftrace/blob/d95a2d14c097bdd5b9de0fdfd216b6c695ed6a2e/src/ast/codegen_llvm.cpp#L1134

Currently, `str()` only accepts a `Type::integer` value. Even if we fix `str()` to accepts a `Type::styring` value, the copy would occur before truncating. So to cope with this problem, special care is needed if a type of the argument of `str()` is `Type::string`.
Alternatively, I think address of operater `&` (#553) is usable to pass an address to `str()`.

**Comment by brad-lewis** [comment]
Thanks @brendangregg and @mmisono.  I understand this is not a bug but that I was misusing str().   Good to hear there are efforts underway to solve the problem.  I'd like to help out if  collaborating on this would be useful.  

What I really am going for here is a latency histogram.   When I use spa_name (Type::string, char [256]) as one of the items in the key:

@latency[$spa->$spa_name] = hist($delta)

 I see two copies on the stack, a local variable and then a key variable: 

  %"@latency_key" = alloca [256 x i8]
  %"$spa_name" = alloca [256 x i8]

Using a map instead of the stack for Type:string sounds very promising.  Would the ongoing effort to put Type::string in the stack affect only the local variable or would the key also be on the map?  If the key is still on the stack then there would be a limit for one map aggregation, e.g. throughput couldn't be aggregated in the same probe.

**Comment by mmisono** [comment]
Storing strings in a map is definitely useful. The issue is #305. It is possible to use a map value directly as a key in Linux v4.18 (https://github.com/torvalds/linux/commit/d71962f3e627b5941804036755c844fabfb65ff5).) So although #305 does not consider a map key, it would help to reduce stack consumption regarding map accesses.
(I don't read the threads thoroughly but does it seem some kernel support is needed? More investigation is needed to understand the problem.)

As of two copies, I noticed that this was because bpftrace always allocated a new buffer for a key and then performed memcpy.
https://github.com/iovisor/bpftrace/blob/d95a2d14c097bdd5b9de0fdfd216b6c695ed6a2e/src/ast/codegen_llvm.cpp#L1503-L1516
This is needed to support multiple values as a key (e.g., `@Latency[$spa->spa_name, pid]`), but I believe there is no need to copy if a key is composed of one value.

I'm working on another map issue in #1009. Along with this, I'll try to work to eliminate the memcpy as mentioned above (maybe next week).

**Comment by fbs** [comment]
Yes we should just skip the copy when `lmap.vargs.size() == 1`. I was looking at a way to order our variables in a way that we do not have to copy them all the time but that is quite tricky. 

> bpftrace treats char * and char [] differently, and the type of former is Type::integer and the latter is Type::string [...]
> Currently, str() only accepts a Type::integer value. Even if we fix str() to accepts a Type::styring value, the copy would occur before truncating.

I'm not sure i like `type::string`. The only condition for being a string is being a `char[]`. As a result you cannot index it like you can with an array. Also `uint8_t abc[123]` and `char abc[123]` behave completely different now, the latter is accepted by `printf %s` the former is not, which will surprise/confuse users. Maybe we should have:

- All strings be arrays
- `str` accepts:
  - pointers where the pointee is an 1 byte wide type
  - "char" arrays, as it provides an easy way to read  a substring. E.g. `str($f->myarray, 10)`. 
- `printf %s` must validate that the value is some 1 byte wide type and warn/error
- string literals can stay a special type (type::stringlit) as that will make semantic analysis easier

**Comment by wubo0067** [comment]
How to output the file name?
struct file->f_path.dentry->d_iname，the d_iname type is unsigned char d_iname[DNAME_INLINE_LEN];
